### PR TITLE
Authz template sub messages

### PR DESCRIPTION
**Pull request by mangchiandjjoe**
**What this PR does / why we need it**:

We need a authorization template for mixer adapters

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

authz template has sub messages

**Release note**:

```release-note
Authorization adapter template.
Any authorization module needs to answer the question: Is a “subject” allowed to do an “action”. 
```

### CODE REVIEW

File: mixer/template/authz/template.proto [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  string service = 2;
+  // The what action is being taken.
+  string method = 3;
+  // The resource within the service.
+  string path = 4;
+  // Additional data about the action for use in policy.
+  map<string, istio.mixer.v1.config.descriptor.ValueType> extra = 5;
+}
+
+message Template {
```
Comment (guptasu) [inline mixer/template/authz/template.proto]: Consider adding more description and example config. sample : https://github.com/istio/istio/blob/master/mixer/template/listentry/template.proto#L23

File: mixer/template/authz/template.proto [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import "mixer/v1/config/descriptor/value_type.proto";
+import "mixer/v1/template/extensions.proto";
+
+option (istio.mixer.v1.template.template_variety) = TEMPLATE_VARIETY_CHECK;
+
+message Subject {
+  // The user.
+  string user = 1;
+  // Groups the subject belongs to.
+  repeated string groups = 2;
```
Comment (guptasu) [inline mixer/template/authz/template.proto]: `repeated`. This is the first template we have used repeated. repeated fields are not completely supported yet. I can make it work in 1/2 day (just need to add its procession in the bootstraping code (loading yaml and generating instance objects), the code for generating artifacts to be used by the adapter is already present). 

This should not block your adapter implementation thought. But unfortunately without change from me, you won't be able to test your adapter e2e with Mixer.

**Comment by mangchiandjjoe** [comment]
I created a design document for the authz template https://goo.gl/1y3Z1v

**Comment by liminw** [comment]
@guptasu Sunny, the template was originally from Istio RBAC[ design doc](https://docs.google.com/document/d/1dKXUEOxrj4TWZKrW7fx_A-nrOdVD4tYolpjgT8DYBTY/edit#heading=h.uy42ya2lsfue). It has been reviewed and approved. 

There is one thing Jae noticed is that, "name" under "subject" is not allowed. The error message says that "name" is a reserved field. Do you think it is a bug? "name" is allowed to be a field name in messages.

**Comment by guptasu** [comment]
It is intentional. We have reserved the name field for instance name value.

File: mixer/template/authz/template.proto (hunk 1 above)
Comment (mangchiandjjoe) [reply mixer/template/authz/template.proto]: Done

File: mixer/template/authz/template.proto (hunk 2 above)
Comment (mangchiandjjoe) [reply mixer/template/authz/template.proto]: That should be fine. We can keep this field repeated for now and write a e2e test after the implementation.

File: mixer/template/authz/template.proto [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  // Additional attributes about the subject.
+  map<string, istio.mixer.v1.config.descriptor.ValueType> extra = 3;
+}
+
+message Action {
```
Comment (guptasu) [inline mixer/template/authz/template.proto]: Add message comment

File: mixer/template/authz/template.proto [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+syntax = "proto3";
+
+package authz;
+
+import "mixer/v1/config/descriptor/value_type.proto";
+import "mixer/v1/template/extensions.proto";
+
+option (istio.mixer.v1.template.template_variety) = TEMPLATE_VARIETY_CHECK;
+
+message Subject {
```
Comment (guptasu) [inline mixer/template/authz/template.proto]: Add message comment.

File: mixer/template/authz/template.proto [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+syntax = "proto3";
+
+package authz;
+
+import "mixer/v1/config/descriptor/value_type.proto";
+import "mixer/v1/template/extensions.proto";
+
+option (istio.mixer.v1.template.template_variety) = TEMPLATE_VARIETY_CHECK;
+
```
Comment (guptasu) [inline mixer/template/authz/template.proto]: The template looks good from the structure point of view. For the field names and comments, I will let others (@liminw and other auth experts) to comment on how intuitive they are.

File: mixer/template/authz/template.proto (hunk 3 above)
Comment (mangchiandjjoe) [reply mixer/template/authz/template.proto]: Done

File: mixer/template/authz/template.proto (hunk 4 above)
Comment (mangchiandjjoe) [reply mixer/template/authz/template.proto]: Done

File: mixer/template/authz/template.proto (hunk 5 above)
Comment (mangchiandjjoe) [reply mixer/template/authz/template.proto]: Field names and comments are from the Istio Native RBAC Design V2 https://goo.gl/kof8o2 which was approved

**Comment by mangchiandjjoe** [comment]
/assign @geeknoid 
Hello Martin, could you please take a look at the PR?

File: mixer/template/authz/template.proto [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+// List of subjects
+message Subject {
+  // The user.
+  string user = 1;
+  // Groups the subject belongs to.
```
Comment (geeknoid) [inline mixer/template/authz/template.proto]: What are the valid groups, where are they defined? Can you clarify in the comments?

File: mixer/template/authz/template.proto [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package authz;
```
Comment (geeknoid) [inline mixer/template/authz/template.proto]: Could we replace all uses of authz with authorization?

File: mixer/template/authz/template.proto [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+message Subject {
+  // The user.
+  string user = 1;
+  // Groups the subject belongs to.
+  repeated string groups = 2;
+  // Additional attributes about the subject.
+  map<string, istio.mixer.v1.config.descriptor.ValueType> extra = 3;
+}
+
+// List of actions
```
Comment (geeknoid) [inline mixer/template/authz/template.proto]: This is not a list, it's just one single action.

Could you provide a comment explaining what this is for?

File: mixer/template/authz/template.proto [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+syntax = "proto3";
+
+package authz;
+
+import "mixer/v1/config/descriptor/value_type.proto";
+import "mixer/v1/template/extensions.proto";
+
+option (istio.mixer.v1.template.template_variety) = TEMPLATE_VARIETY_CHECK;
+
+// List of subjects
```
Comment (geeknoid) [inline mixer/template/authz/template.proto]: This is not a list, it's a single subject.

Could you provide a comment explaining what this is for?

File: mixer/template/authz/template.proto [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+  // Additional attributes about the subject.
+  map<string, istio.mixer.v1.config.descriptor.ValueType> extra = 3;
+}
+
+// List of actions
+message Action {
+  // Namespace the target action is taking place in.
+  string namespace = 1;
+  // The Service the action is being taken on.
+  string service = 2;
```
Comment (geeknoid) [inline mixer/template/authz/template.proto]: @douglas-reid @mandarjog 

Is this the right way to reference a service?

Comment (geeknoid) [inline mixer/template/authz/template.proto]: I don't understand the context in which this information is needed. By the time the adapter gets a request, the notion of 'service' has really been erased from the data. Consider that when we report a metric, the metric object doesn't indicate the service the metric is associated with.

File: mixer/template/authz/template.proto [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+  // The what action is being taken.
+  string method = 3;
```
Comment (geeknoid) [inline mixer/template/authz/template.proto]: This is not related to an API method, right? Should this be 'operation' instead?

File: mixer/template/authz/template.proto [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+  // The resource within the service.
+  string path = 4;
+  // Additional data about the action for use in policy.
+  map<string, istio.mixer.v1.config.descriptor.ValueType> extra = 5;
+}
+
+// Authz template defines subject and action. Adapters using authz template
```
Comment (geeknoid) [inline mixer/template/authz/template.proto]: Adapters "implement" a template, they don't "use" a template.

File: mixer/template/authz/template.proto [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+// should validate “subject” allowed to do an “action”.
```
Comment (geeknoid) [inline mixer/template/authz/template.proto]: should validate that the "subject" is allowed to do the given "action"

File: mixer/template/authz/template.proto [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+//
+// When writing the configuration, the value for the fields associated with this template can either be a
```
Comment (geeknoid) [inline mixer/template/authz/template.proto]: Let's remove this paragraph, that's how templates work in general and we don't need to explain that here.

File: mixer/template/authz/template.proto [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+// must match the datatype of the field.
+//
+// Example config:
+// ```
+// apiVersion: "config.istio.io/v1alpha2"
+// kind: authz
+// metadata:
+//   name: authinfo
+//   namespace: istio-system
+// Spec:
```
Comment (geeknoid) [inline mixer/template/authz/template.proto]: Lowercase s

File: mixer/template/authz/template.proto [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+//  subject:
+//    user: source.user | request.auth.token[user] | ""
+//    groups: request.auth.token[groups]
+//    extra:
+//     iss: request.auth.token["iss"]
+//  action:
```
Comment (geeknoid) [inline mixer/template/authz/template.proto]: I worry about reusing the word action here, since that has a generic meaning in Mixer. Maybe "demand"? A subject makes a demand?

@mandarjog

File: mixer/template/authz/template.proto [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+//    namespace: target.namespace | "default"
+//    service: target.service | ""
+//    path: request.path | "/"
+//    method: request.method | "post"
+//    extra:
+//      version: destination.labels[version] | ""
+//  ```
+message Template {
+  Subject subject = 1;
```
Comment (geeknoid) [inline mixer/template/authz/template.proto]: Comments on the fields please.

File: mixer/template/authz/template.proto [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+option (istio.mixer.v1.template.template_variety) = TEMPLATE_VARIETY_CHECK;
+
+// List of subjects
+message Subject {
+  // The user.
+  string user = 1;
+  // Groups the subject belongs to.
+  repeated string groups = 2;
+  // Additional attributes about the subject.
+  map<string, istio.mixer.v1.config.descriptor.ValueType> extra = 3;
```
Comment (douglas-reid) [inline mixer/template/authz/template.proto]: `extra` is not a great name, imho. can we use something like `attributes` or even something like `metadata` instead?  It might help to have a better comment on `Subject` that explains what a subject is (that doesn't use the word subject in the description) and what `extra` you might need/want for a subject.

File: mixer/template/authz/template.proto (hunk 10 above)
Comment (douglas-reid) [reply mixer/template/authz/template.proto]: is this the `destination.service` ?

File: mixer/template/authz/template.proto [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+  // Namespace the target action is taking place in.
+  string namespace = 1;
+  // The Service the action is being taken on.
+  string service = 2;
+  // The what action is being taken.
+  string method = 3;
+  // The resource within the service.
+  string path = 4;
+  // Additional data about the action for use in policy.
+  map<string, istio.mixer.v1.config.descriptor.ValueType> extra = 5;
```
Comment (douglas-reid) [inline mixer/template/authz/template.proto]: again, i think we want a more descriptive name than `extra`.  Is this stuff metadata about the action?

File: mixer/template/authz/template.proto (hunk 13 above)
Comment (douglas-reid) [reply mixer/template/authz/template.proto]: is allowed to *perform* the specified action ?

File: mixer/template/authz/template.proto (hunk 12 above)
Comment (douglas-reid) [inline mixer/template/authz/template.proto]: Can we add more detail about the purpose here? This will end up in docs on the website, and while authorization is reasonably clear to sophisticated users, I think we could benefit from some lingo clarity.  Maybe something like:

> The authorization template defines parameters for performing policy enforcement within Istio. It is primarily concerned with enabling Mixer adapters to make decisions about *who* is allowed to do *what*. In this template, the *who* is defined in a *Subject* message. The *what* is defined in the *Action* message. During a Mixer Check call, these values will be populated based on configuration from request attributes and passed to individual authorization adapters to adjudicate.

You can probably describe it better, but I think that would be helpful (and maybe eliminate the need for the following paragraph).

File: mixer/template/authz/template.proto (hunk 18 above)
Comment (liminw) [reply mixer/template/authz/template.proto]: It was previously named "attributes". It was renamed because Martin thought that "attributes" conflicts with Istio terms of "attributes". It is not "metadata" either. This field represents the custom attributes (in addition to "user" and "groups") that can be used to identify the subject. It is hard to come up with a good name. Do you have other suggestions for the name?

File: mixer/template/authz/template.proto (hunk 8 above)
Comment (liminw) [reply mixer/template/authz/template.proto]: Action defines "how the resource is accessed".

File: mixer/template/authz/template.proto (hunk 10 above)
Comment (liminw) [reply mixer/template/authz/template.proto]: This will only be used in a Check request. "service" refers to destination service name.

File: mixer/template/authz/template.proto (hunk 11 above)
Comment (liminw) [reply mixer/template/authz/template.proto]: It can be a HTTP method ("GET", "POST", etc.) or gRPC method.

File: mixer/template/authz/template.proto (hunk 18 above)
Comment (geeknoid) [reply mixer/template/authz/template.proto]: "properties"?

File: mixer/template/authz/template.proto (hunk 7 above)
Comment (mangchiandjjoe) [reply mixer/template/authz/template.proto]: Renamed

File: mixer/template/authz/template.proto (hunk 9 above)
Comment (mangchiandjjoe) [reply mixer/template/authz/template.proto]: Updated

File: mixer/template/authz/template.proto (hunk 6 above)
Comment (mangchiandjjoe) [reply mixer/template/authz/template.proto]: Updated comment

File: mixer/template/authz/template.proto (hunk 8 above)
Comment (mangchiandjjoe) [reply mixer/template/authz/template.proto]: Updated

File: mixer/template/authz/template.proto (hunk 12 above)
Comment (mangchiandjjoe) [reply mixer/template/authz/template.proto]: Updated

File: mixer/template/authz/template.proto (hunk 13 above)
Comment (mangchiandjjoe) [reply mixer/template/authz/template.proto]: Updated

File: mixer/template/authz/template.proto (hunk 14 above)
Comment (mangchiandjjoe) [reply mixer/template/authz/template.proto]: Done

File: mixer/template/authz/template.proto (hunk 15 above)
Comment (mangchiandjjoe) [reply mixer/template/authz/template.proto]: Done

File: mixer/template/authz/template.proto (hunk 17 above)
Comment (mangchiandjjoe) [reply mixer/template/authz/template.proto]: Done

File: mixer/template/authz/template.proto (hunk 12 above)
Comment (mangchiandjjoe) [reply mixer/template/authz/template.proto]: Replaced

File: mixer/template/authorization/template.proto [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+syntax = "proto3";
+
+package authorization;
+
+import "mixer/v1/config/descriptor/value_type.proto";
+import "mixer/v1/template/extensions.proto";
+
+option (istio.mixer.v1.template.template_variety) = TEMPLATE_VARIETY_CHECK;
+
+// A set of subjects contains a list of attributes that identify
```
Comment (geeknoid) [inline mixer/template/authorization/template.proto]: "A set of subjects" -> A subject

File: mixer/template/authorization/template.proto [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+message Subject {
+  // The user.
+  string user = 1;
+  // Groups the subject belongs to depending on the authentication mechanism,
+  // "groups" are normally populated from JWT claim or client certificate.
+  // The operator can define how it is populated when creating an instance of
+  // the template.
+  repeated string groups = 2;
+  // Additional attributes about the subject.
+  map<string, istio.mixer.v1.config.descriptor.ValueType> extra = 3;
```
Comment (geeknoid) [inline mixer/template/authorization/template.proto]: properties?

File: mixer/template/authorization/template.proto [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+
+import "mixer/v1/config/descriptor/value_type.proto";
+import "mixer/v1/template/extensions.proto";
+
+option (istio.mixer.v1.template.template_variety) = TEMPLATE_VARIETY_CHECK;
+
+// A set of subjects contains a list of attributes that identify
+// the caller identity.
+message Subject {
+  // The user.
```
Comment (geeknoid) [inline mixer/template/authorization/template.proto]: Can we expand this comment?

File: mixer/template/authorization/template.proto [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+  // Groups the subject belongs to depending on the authentication mechanism,
+  // "groups" are normally populated from JWT claim or client certificate.
+  // The operator can define how it is populated when creating an instance of
+  // the template.
+  repeated string groups = 2;
+  // Additional attributes about the subject.
+  map<string, istio.mixer.v1.config.descriptor.ValueType> extra = 3;
+}
+
+// A set of actions defines "how the resource is accessed".
```
Comment (geeknoid) [inline mixer/template/authorization/template.proto]: "A set of actions" -> An action

File: mixer/template/authorization/template.proto [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+message Action {
+  // Namespace the target action is taking place in.
+  string namespace = 1;
+  // The Service the action is being taken on.
+  string service = 2;
+  // The what action is being taken.
```
Comment (geeknoid) [inline mixer/template/authorization/template.proto]: The what -> What

File: mixer/template/authorization/template.proto [hunk 25]
```diff
@@ -0,0 +1,86 @@
...
+  string path = 4;
+  // Additional data about the action for use in policy.
+  map<string, istio.mixer.v1.config.descriptor.ValueType> extra = 5;
+}
+
+// The authorization template defines parameters for performing policy
+// enforcement within Istio. It is primarily concerned with enabling Mixer
+// adapters to make decisions about who is allowed to do what.
+// In this template, the who is defined in a Subject message. The what is
+// defined in the Action message. During a Mixer Check call, these values
```
Comment (geeknoid) [inline mixer/template/authorization/template.proto]: the Action -> an Action

File: mixer/template/authorization/template.proto [hunk 26]
```diff
@@ -0,0 +1,86 @@
...
+//  action:
+//    namespace: target.namespace | "default"
+//    service: target.service | ""
+//    path: request.path | "/"
+//    method: request.method | "post"
+//    extra:
+//      version: destination.labels[version] | ""
+//  ```
+message Template {
+  // A set of subjects
```
Comment (geeknoid) [inline mixer/template/authorization/template.proto]: This is not a set. Please rephrase both comments to be more descriptive, you can steal some text from the message-level comment above.

File: mixer/template/authorization/template.proto (hunk 25 above)
Comment (mangchiandjjoe) [reply mixer/template/authorization/template.proto]: Done

File: mixer/template/authorization/template.proto (hunk 24 above)
Comment (mangchiandjjoe) [reply mixer/template/authorization/template.proto]: Done

File: mixer/template/authorization/template.proto (hunk 23 above)
Comment (mangchiandjjoe) [reply mixer/template/authorization/template.proto]: Done

File: mixer/template/authorization/template.proto (hunk 21 above)
Comment (mangchiandjjoe) [reply mixer/template/authorization/template.proto]: Done

File: mixer/template/authorization/template.proto (hunk 22 above)
Comment (mangchiandjjoe) [reply mixer/template/authorization/template.proto]: Update

File: mixer/template/authorization/template.proto (hunk 26 above)
Comment (mangchiandjjoe) [reply mixer/template/authorization/template.proto]: Update

File: mixer/template/authorization/template.proto [hunk 27]
```diff
@@ -0,0 +1,87 @@
...
+//    user: source.user | request.auth.token[user] | ""
+//    groups: request.auth.token[groups]
+//    extra:
+//     iss: request.auth.token["iss"]
+//  action:
+//    namespace: target.namespace | "default"
+//    service: target.service | ""
+//    path: request.path | "/"
+//    method: request.method | "post"
+//    extra:
```
Comment (geeknoid) [inline mixer/template/authorization/template.proto]: extra -> properties

File: mixer/template/authorization/template.proto [hunk 28]
```diff
@@ -0,0 +1,87 @@
...
+// apiVersion: "config.istio.io/v1alpha2"
+// kind: authorization
+// metadata:
+//   name: authinfo
+//   namespace: istio-system
+// spec:
+//  subject:
+//    user: source.user | request.auth.token[user] | ""
+//    groups: request.auth.token[groups]
+//    extra:
```
Comment (geeknoid) [inline mixer/template/authorization/template.proto]: extra -> properties

File: mixer/template/authorization/template.proto [hunk 29]
```diff
@@ -0,0 +1,87 @@
...
+// An action defines "how a resource is accessed".
+message Action {
+  // Namespace the target action is taking place in.
+  string namespace = 1;
+  // The Service the action is being taken on.
+  string service = 2;
+  // What action is being taken.
+  string method = 3;
+  // The resource within the service.
+  string path = 4;
```
Comment (geeknoid) [inline mixer/template/authorization/template.proto]: Should this be called resource? or resource_path?

Comment (mangchiandjjoe) [reply mixer/template/authorization/template.proto]: It is the HTTP REST path within the service. Updated comment.

File: mixer/template/authorization/template.proto (hunk 28 above)
Comment (mangchiandjjoe) [reply mixer/template/authorization/template.proto]: Done

File: mixer/template/authorization/template.proto (hunk 27 above)
Comment (mangchiandjjoe) [reply mixer/template/authorization/template.proto]: Done
